detection_rules/rule_validators.py

```diff
@@ -81,34 +97,44 @@ def validate(self, data: 'QueryRuleData', meta: RuleMeta) -> None:
             return
 
         indexes = data.index or []
-        beats_version = meta.beats_version or beats.get_max_version()
-        ecs_versions = meta.ecs_versions or [ecs.get_max_version()]
+        stack_versions = get_stack_schemas(meta.min_stack_version)
+
+        if not stack_versions:
+            # add latest
+            stack_versions = {CURRENT_PACKAGE: {'beats': beats.get_max_version(), 'ecs': ecs.get_max_version()}}
+
+        if meta.query_validate_master:
+            stack_versions['latest'] = {'beats': 'master', 'ecs': 'master'}
 
         # TODO: remove once py-eql supports ipv6 for cidrmatch
         # Or, unregister the cidrMatch function and replace it with one that doesn't validate against strict IPv4
         with eql.parser.elasticsearch_syntax, eql.parser.ignore_missing_functions:
             parsed = eql.parse_query(self.query)
 
         beat_types = [index.split("-")[0] for index in indexes if "beat-*" in index]
-        beat_schema = beats.get_schema_from_eql(parsed, beat_types, version=beats_version) if beat_types else None
 
-        for version in ecs_versions:
-            schema = ecs.get_kql_schema(indexes=indexes, beat_schema=beat_schema, version=version)
+        for stack_version, mapping in stack_versions.items():
+            beats_version = mapping['beats']
+            ecs_version = mapping['ecs']
+            err_trailer = f'stack: {stack_version}, beats: {beats_version}, ecs: {ecs_version}'
+
+            beat_schema = beats.get_schema_from_eql(parsed, beat_types, version=beats_version) if beat_types else None
+
+            schema = ecs.get_kql_schema(indexes=indexes, beat_schema=beat_schema, version=ecs_version)
```

Comment:
These could probably a helper method on the base too:
`def get_kql_schema(self, indexes, beats_version, ecs_version) -> (KqlSchema)`

Reply:
4c54015